lib/helpers/signed-aws-es-fetch.js

```diff
@@ -0,0 +1,109 @@
+const aws4 = require('aws4');
+const nodeFetch = require('node-fetch');
+const logger = require('@financial-times/n-logger').default;
+const resolveCname = require('util').promisify(require('dns').resolveCname);
+const memoize = require('lodash/memoize');
+
+const memoizedResolveCname = memoize(resolveCname);
+
+const envKeys = {
+	awsKeys: ['ES_AWS_ACCESS_KEY', 'AWS_ACCESS_KEY', 'AWS_ACCESS_KEY_ID'],
+	awsSecretKeys: ['ES_AWS_SECRET_ACCESS_KEY', 'AWS_SECRET_ACCESS_KEY'],
+	tokenKeys: ['ES_AWS_SESSION_TOKEN', 'AWS_SESSION_TOKEN']
+};
+
+module.exports = async function (url, opts, creds) {
+	opts = opts || {};
+	url = await resolveUrlAndHost(url);
+	return signedFetch(url, opts, creds);
+};
+
+function getURL (url) {
+	try {
+		const urlObject = new URL(url);
+		return urlObject;
+	} catch (error) {
+		logger.error({
+			event: 'Invalid URL',
+			data: {
+				url
+			}
+		});
+		throw error;
+	}
+}
+
+async function resolveValidHost (urlObject) {
+	const hosts = await memoizedResolveCname(urlObject.host);
+	if (hosts === undefined) {
+		logger.error({
+			event: 'Invalid Host',
+			data: {
+				url: urlObject
+			}
+		});
+		throw new Error('Invalid Host');
+	}
+	return hosts[0];
+}
+
+async function resolveUrlAndHost (url) {
+	const urlObject = getURL(url);
+	if (
+		!/\.es\.amazonaws\.com$/.test(urlObject.host) &&
+		!process.env.AWS_SIGNED_FETCH_DISABLE_DNS_RESOLUTION
+	) {
+		const host = await resolveValidHost(urlObject);
+		url = url.replace(urlObject.host, host);
+	}
+	return url;
+}
+
+function getEnv (keys) {
+	const keyName = envKeys[keys].find((keyName) => process.env[keyName]);
+	return process.env[keyName];
+}
+
+function setAwsCredentials (creds) {
```

Comment:
**nitpick:** Should this be something like `defaultAwsCredentials`? The naming confused me a little when I was trying to unpick the logic.